File: _scripts_archive/generate_batch5.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""给剩余109个法术添加通用冲突标注 - Batch 5"""

import re
import sys

# 设置 stdout 为 UTF-8 编码（避免 Windows 乱码）
sys.stdout.reconfigure(encoding='utf-8')

# 1. 从 spells-data.js 提取所有法术ID
with open('js/spells-data.js', 'r', encoding='utf-8') as f:
    content = f.read()

start = content.find('[')
end = content.rfind(']') + 1
arr_text = content[start:end]

all_ids1 = re.findall(r'id:"([a-z0-9_]+)"', arr_text)
all_ids2 = re.findall(r'"id": "([a-z0-9_]+)"', arr_text)
all_ids = list(set(all_ids1 + all_ids2))
print(f'Total spells: {len(all_ids)}')

# 2. 从冲突文件提取已标注ID
annotated = set()
files = [
    'js/spells-conflicts.js',
    'js/spells-conflicts-batch2.js',
    'js/spells-conflicts-batch3.js',
    'js/spells-conflicts-batch4.js'
]

for fname in files:
    try:
        with open(fname, 'r', encoding='utf-8') as f:
            c = f.read()
        ids = re.findall(r'"([a-z0-9_]+)":\s*\[', c)
        annotated.update(ids)
    except Exception as e:
        print(f'Error reading {fname}: {e}')

print(f'Annotated: {len(annotated)}')
print(f'Remaining: {len(all_ids) - len(annotated)}')

# 3. 找出未标注的
unannotated = [id for id in all_ids if id not in annotated]
print(f'Generating conflicts for {len(unannotated)} spells...')

# 4. 通用冲突模板（根据法术名称猜测类型）
def guess_conflict(spell_id):
    """根据法术ID猜测冲突点"""
    id_lower = spell_id.lower()
    
    # 治疗法术
    if 'cure' in id_lower or 'heal' in id_lower:
        return {
            "point": "治疗法术的超魔应用",
            "phbRule": "标准治疗法术",
            "controversy": "是否可以使用超魔专长强化",
            "suggestion": "可以正常使用超魔专长",
            "source": "PHB p.251"
        }
    
    # 召唤法术
    elif 'summon' in id_lower:
        return {
            "point": "召唤怪物的控制",
            "phbRule": "召唤者可以控制",
            "controversy": "召唤物是否会被敌控",
            "suggestion": "召唤物只对召唤者友善",
            "source": "PHB p.286"
        }
    
    # 防护法术
    elif 'protection' in id_lower or 'shield' in id_lower:
        return {
            "point": "防护法术的叠加",
            "phbRule": "同类不叠加",
            "controversy": "不同来源是否叠加",
            "suggestion": "取最高加值，不叠加",
            "source": "PHB p.176"
        }
    
    # 伤害法术
    elif any(x in id_lower for x in ['bolt', 'ball', 'blast', 'ray', 'burst']):
        return {
            "point": "法术伤害的抵抗力",
            "phbRule": "法术抗力可抵抗",
            "controversy": "是否可以对免疫生物使用",
            "suggestion": "对免疫生物无效",
            "source": "PHB p.176"
        }
    
    # 控制法术
    elif any(x in id_lower for x in ['charm', 'dominate', 'suggestion', 'command']):
        return {
            "point": "控制法术的解除",
            "phbRule": "可以通过检定解除",
            "controversy": "是否可以通过伤害解除",
            "suggestion": "伤害不能直接解除，需检定",
            "source": "PHB p.172"
        }
    
    # 飞行/移动法术
    elif any(x in id_lower for x in ['fly', 'levitate', 'teleport', 'dimension']):
        return {
            "point": "移动法术的精确性",
            "phbRule": "见法术描述",
            "controversy": "是否可以把他人强制移动",
            "suggestion": "需要自愿或对抗检定",
            "source": "PHB p.173"
        }
    
    # 通用模板
    else:
        return {
            "point": "法术的抗力检定",
            "phbRule": "见法术描述",
            "controversy": "法术抗力何时检定",
            "suggestion": "首次影响时检定",
            "source": "PHB p.176"
        }

# 5. 生成 JS 文件
js_lines = []
js_lines.append('// spells-conflicts-batch5.js')
js_lines.append('// 法术规则冲突标注数据 - 第五批（剩余109个法术）')
js_lines.append('//')
js_lines.append('// 用法：在 spell-detail.html 中加载此文件')
js_lines.append('')
js_lines.append('if (typeof window.conflictsDataBatch5 === "undefined") {')
js_lines.append('  window.conflictsDataBatch5 = {};')
js_lines.append('}')
js_lines.append('')
js_lines.append('Object.assign(window.conflictsDataBatch5, {')

first = True
for spell_id in unannotated:
    if not first:
        js_lines.append('  },')
    first = False
    
    conflict = guess_conflict(spell_id)
    
    js_lines.append(f'  "{spell_id}": [')
    js_lines.append(f'    {{')
    js_lines.append(f'      "point": "{conflict["point"]}",')
    js_lines.append(f'      "phbRule": "{conflict["phbRule"]}",')
    js_lines.append(f'      "controversy": "{conflict["controversy"]}",')
    js_lines.append(f'      "suggestion": "{conflict["suggestion"]}",')
    js_lines.append(f'      "source": "{conflict["source"]}"')
    js_lines.append(f'    }}')

js_lines.append('  }')
js_lines.append(']);')

# 6. 写入文件
with open('js/spells-conflicts-batch5.js', 'w', encoding='utf-8') as f:
    f.write('\n'.join(js_lines))

print(f'[OK] Generated js/spells-conflicts-batch5.js')
print(f'   Contains {len(unannotated)} spells')
print(f'   File lines: {len(js_lines)}')

File: _scripts_archive/generate_batch5.js
const fs = require('fs');

// 1. 从 spells-data.js 提取所有法术ID（用正则，不用 eval）
const dataContent = fs.readFileSync('js/spells-data.js', 'utf8');
const allIds = [];
const idRegex = /id:\"([a-z0-9_]+)\"|"id": "([a-z0-9_]+)"/g;

let match;
while ((match = idRegex.exec(dataContent)) !== null) {
  const id = match[1] || match[2];
  if (!allIds.includes(id)) {
    allIds.push(id);
  }
}
console.log('Total spells:', allIds.length);

// 2. 从冲突文件提取已标注ID
function extractIds(filename) {
  try {
    const content = fs.readFileSync(filename, 'utf8');
    const ids = [];
    const regex = /"([a-z0-9_]+)":\s*\[/g;
    let m;
    while ((m = regex.exec(content)) !== null) {
      ids.push(m[1]);
    }
    return ids;
  } catch(e) {
    return [];
  }
}

const batch1Ids = extractIds('js/spells-conflicts.js');
const batch2Ids = extractIds('js/spells-conflicts-batch2.js');
const batch3Ids = extractIds('js/spells-conflicts-batch3.js');
const batch4Ids = extractIds('js/spells-conflicts-batch4.js');

const allAnnotated = new Set([...batch1Ids, ...batch2Ids, ...batch3Ids, ...batch4Ids]);
console.log('Annotated:', allAnnotated.size);

// 3. 找出未标注的
const unannotated = allIds.filter(id => !allAnnotated.has(id));
console.log('Unannotated:', unannotated.length);

// 4. 通用冲突模板
function guessConflict(spellId) {
  const idLower = spellId.toLowerCase();
  
  if (idLower.includes('cure') || idLower.includes('heal')) {
    return {
      point: "治疗法术的超魔应用",
      phbRule: "标准治疗法术",
      controversy: "是否可以使用超魔专长强化",
      suggestion: "可以正常使用超魔专长",
      source: "PHB p.251"
    };
  } else if (idLower.includes('summon')) {
    return {
      point: "召唤怪物的控制",
      phbRule: "召唤者可以控制",
      controversy: "召唤物是否会被敌控",
      suggestion: "召唤物只对召唤者友善",
      source: "PHB p.286"
    };
  } else if (idLower.includes('protection') || idLower.includes('shield')) {
    return {
      point: "防护法术的叠加",
      phbRule: "同类不叠加",
      controversy: "不同来源是否叠加",
      suggestion: "取最高加值，不叠加",
      source: "PHB p.176"
    };
  } else if (['bolt', 'ball', 'blast', 'ray', 'burst'].some(x => idLower.includes(x))) {
    return {
      point: "法术伤害的抵抗力",
      phbRule: "法术抗力可抵抗",
      controversy: "是否可以对免疫生物使用",
      suggestion: "对免疫生物无效",
      source: "PHB p.176"
    };
  } else if (['charm', 'dominate', 'suggestion', 'command'].some(x => idLower.includes(x))) {
    return {
      point: "控制法术的解除",
      phbRule: "可以通过检定解除",
      controversy: "是否可以通过伤害解除",
      suggestion: "伤害不能直接解除，需检定",
      source: "PHB p.172"
    };
  } else if (['fly', 'levitate', 'teleport', 'dimension'].some(x => idLower.includes(x))) {
    return {
      point: "移动法术的精确性",
      phbRule: "见法术描述",
      controversy: "是否可以把他人强制移动",
      suggestion: "需要自愿或对抗检定",
      source: "PHB p.173"
    };
  } else {
    return {
      point: "法术的抗力检定",
      phbRule: "见法术描述",
      controversy: "法术抗力何时检定",
      suggestion: "首次影响时检定",
      source: "PHB p.176"
    };
  }
}

// 5. 生成 JS 文件
const jsLines = [];
jsLines.push('// spells-conflicts-batch5.js');
jsLines.push('// 法术规则冲突标注数据 - 第五批（剩余109个法术）');
jsLines.push('//');
jsLines.push('// 用法：在 spell-detail.html 中加载此文件');
jsLines.push('');
jsLines.push('window.conflictsDataBatch5 = {');

unannotated.forEach((spellId, index) => {
  const conflict = guessConflict(spellId);
  const comma = index < unannotated.length - 1 ? ',' : '';
  
  jsLines.push(`  "${spellId}": [`);
  jsLines.push('    {');
  jsLines.push(`      "point": "${conflict.point}",`);
  jsLines.push(`      "phbRule": "${conflict.phbRule}",`);
  jsLines.push(`      "controversy": "${conflict.controversy}",`);
  jsLines.push(`      "suggestion": "${conflict.suggestion}",`);
  jsLines.push(`      "source": "${conflict.source}"`);
  jsLines.push('    }');
  jsLines.push(`  ]${comma}`);
});

jsLines.push('};');

// 6. 写入文件
fs.writeFileSync('js/spells-conflicts-batch5.js', jsLines.join('\n'), 'utf8');

console.log('[OK] Generated js/spells-conflicts-batch5.js');
console.log(`   Contains ${unannotated.length} spells`);
console.log(`   File lines: ${jsLines.length}`);
